### PR TITLE
Fix #1571 - building.clone breaks space surface matching

### LINKED ISSUE #1571: New building.clone breaks space surface matching
I made a quick measure to test this out. Found a similar issue to what I see if I loop through spaces in an external model and clone them over one by one. The surface matching breaks.

Here is the test measure
https://github.com/NREL/OpenStudio-measures/blob/develop/NREL%20working%20measures/clone_building_from_external_model/measure.rb

Here is a screenshot in SketchUp
4-1-2015 3-09-19 pm

OS app as well for reference.
4-1-2015 3-21-39 pm

@macumber any ideas? I gave this a major flag because we have a lot of different ways people are faking a (change seed model) measure. I was hoping to setup a clean example for others to use.

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

Fix #1571 - building.clone breaks space surface matching

### Pull Request Author

 - [x] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [x] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [x] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/model/Building.cpp [hunk 1]
```diff
@@ -232,6 +235,25 @@ namespace model {
...
+            surfaceToAdjacentMap.emplace(s.nameString(), adjS->nameString());
+          }
+        }
+
+        std::map<std::string, std::string> subSurfaceToAdjacentMap;
+        for (const auto& s : model().getConcreteModelObjects<SubSurface>()) {
+          if (auto adjS = s.adjacentSubSurface()) {
+            subSurfaceToAdjacentMap.emplace(s.nameString(), adjS->nameString());
+          }
+        }
```
Comment (jmarrec) [inline src/model/Building.cpp:255]: Store a map. See comments for why I think it's safe

File: src/model/Building.cpp [hunk 2]
```diff
@@ -248,6 +270,26 @@ namespace model {
...
+        for (auto& s : t_model.getConcreteModelObjects<SubSurface>()) {
+          if (!s.adjacentSubSurface()) {
+            if (auto it = subSurfaceToAdjacentMap.find(s.nameString()); it != subSurfaceToAdjacentMap.end()) {
+              auto adjacentSubSurface = t_model.getConcreteModelObjectByName<SubSurface>(it->second);
+              OS_ASSERT(adjacentSubSurface);
+              s.setAdjacentSubSurface(adjacentSubSurface.get());
+            }
+          }
+        }
```
Comment (jmarrec) [inline src/model/Building.cpp:292]: setAdjacentSurface on the cloned model, based on the name maps.

File: src/model/test/Building_GTest.cpp [hunk 3]
```diff
@@ -787,3 +788,19 @@ TEST_F(ModelFixture, Building_exteriorPerimeter) {
...
+  auto nMatched = std::count_if(surfaces.cbegin(), surfaces.cend(), [](const auto& s) { return s.adjacentSurface(); });
+  EXPECT_EQ(8, nMatched);
+
+  Model m2;
+  auto bClone = b.clone(m2);
+  auto surfaceClones = m2.getConcreteModelObjects<Surface>();
+  EXPECT_EQ(surfaces.size(), surfaceClones.size());
+  auto nMatchedClone = std::count_if(surfaceClones.cbegin(), surfaceClones.cend(), [](const auto& s) { return s.adjacentSurface(); });
+  EXPECT_EQ(nMatched, nMatchedClone);
+}
```
Comment (jmarrec) [inline src/model/test/Building_GTest.cpp:806]: Gtest

**Review by DavidGoldwasser — APPROVED** [review]
Looks good, storing and assigning matches seems like clean approach.

**Comment by kbenne** [comment]
@jmarrec I'm pretty sure this is failing for unrelated reasons. I've looked over it and it seems ready to go, unless you have any concerns?

**Review by kbenne — APPROVED** [review]

**Comment by jmarrec** [comment]
@kbenne looks all green now.

**Pull request merged**